### PR TITLE
Dev joycommon and advmame auto gamepad fixes

### PR DESCRIPTION
**Pull request by Langerz82**
Fixes to:
Advmame - added xbox controller analog to trigger values.
Advmame - fixes to hat values.
Joy Common - Fixes to when /proc/bus/input/devices has js[0-9] at beginning or end of Handlers value.

Reference Issue:
https://github.com/EmuELEC/EmuELEC/issues/827
Joy Common issue appeared when I was testing on one of my boxes.

### CODE REVIEW

**Comment by Langerz82** [comment]
@shantigilbert Hmmm, can you show the output of your devices?
```
cat /proc/bus/input/devices
```

**Comment by shantigilbert** [comment]
> cat /proc/bus/input/devices

```
cat /proc/bus/input/devices
I: Bus=0010 Vendor=1b8e Product=0cec Version=0001
N: Name="cec_input"
P: Phys=
S: Sysfs=/devices/virtual/input/input0
U: Uniq=
H: Handlers=kbd event0 
B: PROP=0
B: EV=3
B: KEY=1 0 0 10000000000000 0

I: Bus=0003 Vendor=1997 Product=1221 Version=0100
N: Name="Riite Mini Keybo"
P: Phys=usb-xhci-hcd.0.auto-1.2/input0
S: Sysfs=/devices/platform/ff500000.dwc3/xhci-hcd.0.auto/usb1/1-1/1-1.2/1-1.2:1.0/0003:1997:1221.0001/input/input1
U: Uniq=
H: Handlers=sysrq kbd leds event1 
B: PROP=0
B: EV=120013
B: KEY=1000000000007 ff800000000007ff febeffdfffefffff fffffffffffffffe
B: MSC=10
B: LED=1f

I: Bus=0003 Vendor=1997 Product=1221 Version=0100
N: Name="Riite Mini Keybo"
P: Phys=usb-xhci-hcd.0.auto-1.2/input1
S: Sysfs=/devices/platform/ff500000.dwc3/xhci-hcd.0.auto/usb1/1-1/1-1.2/1-1.2:1.1/0003:1997:1221.0002/input/input2
U: Uniq=
H: Handlers=kbd mouse0 event2 
B: PROP=0
B: EV=1f
B: KEY=3007f 0 0 483ffff17aff32d bf54444600000000 1f0001 130c730b17c000 267bfad941dfed 9e168000004400 10000002
B: REL=143
B: ABS=100000000
B: MSC=10

I: Bus=0019 Vendor=0000 Product=0000 Version=0000
N: Name="meson-ir"
P: Phys=meson-ir/input0
S: Sysfs=/devices/platform/ff808040.meson-ir/rc/rc0/input4
U: Uniq=
H: Handlers=kbd event4 
B: PROP=0
B: EV=100013
B: KEY=fff 0 4200108fc32e 237605100000000 0 700158000 419240004801 9e96c000000000 10000ffc
B: MSC=10

I: Bus=0003 Vendor=045e Product=028e Version=0110
N: Name="Microsoft X-Box 360 pad"
P: Phys=usb-xhci-hcd.0.auto-1.4/input0
S: Sysfs=/devices/platform/ff500000.dwc3/xhci-hcd.0.auto/usb1/1-1/1-1.4/1-1.4:1.0/input/input6
U: Uniq=
H: Handlers=event3 js0 
B: PROP=0
B: EV=20000b
B: KEY=7cdb000000000000 0 0 0 0
B: ABS=3003f
B: FF=107030000 0
```

**Comment by shantigilbert** [comment]
Actually it configures 9 players to the same gamepad

**Comment by Langerz82** [comment]
Woah lol, ok I'll make a change that hopefully fixes the issue.

**Comment by Langerz82** [comment]
@shantigilbert try this new change, it should work ok. The problem with the prev regex was that it was detecting incorrect values.

**Comment by shantigilbert** [comment]
there is also an issue with gamepad names: e.g. 

Name: `HORI CO.,LTD. HORIPAD S`

your script: `hori_co_ltd_horipad_s`
it should be: `hori_coltd_horipad_s`

**Comment by shantigilbert** [comment]
this is how I used to change the name: 

https://github.com/EmuELEC/EmuELEC/blob/dev_4.3/packages/sx05re/emuelec/bin/set_advmame_joy.sh#L111

**Comment by Langerz82** [comment]
hey I copy pasted your code into the new advmame.

**Comment by shantigilbert** [comment]
> @shantigilbert try this new change, it should work ok. The problem with the prev regex was that it was detecting incorrect values.

the commit you posted here now does not even detect the gamepad :(

**Comment by shantigilbert** [comment]
I think the issue is with the function `jc_get_players()` it just goes 8 times ignoring whether the players are set in es_settings.cfg or not

I tried following the script so I could debug it, but its just too complicated. 

Can you explain what is the final purpose of this script change?

maybe you could search for how many (js) devices are actually connected in /dev/input and then check the order in es_settings.cfg after that

Right now you are doing 8 passes every time, even if only 1 controller is connected.  `  for ((y = 1; y <= 8; y++)); do` and you are not clearing the variables, hence why we end up with 8 players.

**Comment by Langerz82** [comment]
@shantigilbert Ok, let it with me and I'll properly test with a couple different controllers at the same time and submit a fix.

**Comment by shantigilbert** [comment]
also, for more info: 

```
BTN_CFG=0 1 2 3 4 5 6 7
Getting GUID for INPUT P1GUID
CONTROLLER: 1 js0 030000000d0f0000c100000011010000 HORI CO.,LTD. HORIPAD S
GC_CONFIG=030000000d0f0000c100000011010000,HORI CO. LTD. HORIPAD S,a:b1,b:b2,back:b8,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b13,leftshoulder:b4,leftstick:b10,lefttrigger:b6,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b11,righttrigger:b7,rightx:a2,righty:a3,start:b9,x:b0,y:b3,platform:Linux,
Invalid argument 'joystick_digital[hori_coltd_horipad_s,stick4,y,up] or joystick_digital[hori_coltd_horipad_s,stick,y,up]' for option 'input_map[p1_up]'.
Invalid stick 'stick4'.
Valid format is keyboard[BOARD,KEY]/joystick_button[JOYSTICK,BUTTON]/mouse_button[MOUSE,BUTTON].
```

**Comment by Langerz82** [comment]
For advmame I left out the hat config, unless the user adds in emuelec.conf
```
advmame_add_hat=1
advmame_btn_h0=4 # It may be different for you, go in advance mame and set the dpad to a button to get correct index.
```
Concerning the iterations, it should not loop through as it has continue statements which returns the iteration to the top again.
I looked at another way of doing it, but it is more complex than it needs to be.

I tested with your input too and it worked for me.

**Comment by shantigilbert** [comment]
> For advmame I left out the hat config, unless the user adds in emuelec.conf

I will send you a utility to determine this value, so that users don't have to guess it. I modified `advj` so that it returns the gamepad information that advmame needs for example:

My Logitech F310 which needs `4` as the H0

```
advj
Driver raw, joysticks 1
sticks 4 
joy 0 'logitech_gamepad_f310' 'Logitech Gamepad F310', controls 4, buttons 11, ball axes 0
	control 0 [stick], axes 2
		axe 0 [x]
		axe 1 [y]
	control 1 [stick2], axes 2
		axe 0 [x]
		axe 1 [y]
	control 2 [stick3], axes 2
		axe 0 [x]
		axe 1 [y]
	control 3 [stick4], axes 2
		axe 0 [x]
		axe 1 [y]
	button 0 [button1]
	button 1 [button2]
	button 2 [button3]
	button 3 [button4]
	button 4 [button5]
	button 5 [button6]
	button 6 [button7]
	button 7 [button8]
	button 8 [button9]
	button 9 [button10]
	button 10 [button11]
```
My Hori pad which uses H0=3

```
advj
Driver raw, joysticks 1
sticks 3 
joy 0 'hori_coltd_horipad_s' 'HORI CO.,LTD. HORIPAD S', controls 3, buttons 14, ball axes 0
	control 0 [stick], axes 2
		axe 0 [x]
		axe 1 [y]
	control 1 [stick2], axes 2
		axe 0 [x]
		axe 1 [y]
	control 2 [stick3], axes 2
		axe 0 [x]
		axe 1 [y]
	button 0 [button1]
	button 1 [button2]
	button 2 [button3]
	button 3 [button4]
	button 4 [button5]
	button 5 [button6]
	button 6 [button7]
	button 7 [button8]
	button 8 [button9]
	button 9 [button10]
	button 10 [button11]
	button 11 [button12]
	button 12 [button13]
	button 13 [button14]
```

> Concerning the iterations, it should not loop through as it has continue statements which returns the iteration to the top again.
I looked at another way of doing it, but it is more complex than it needs to be.

you are right, it should not, but the problem is that it is doing it :(

**Comment by shantigilbert** [comment]
[advj.zip](https://github.com/EmuELEC/EmuELEC/files/8187021/advj.zip)
This will be included in EmuELEC and you can use it like this:

`advmame_btn_h0=$(advj | grep sticks: | sed "s|sticks:\ ||"`)

which in theory should return the correct dpad "stick" but of course it needs more testing as I only tested it with a few gamepads

**Comment by shantigilbert** [comment]
Your last 2 commits broke hats, IMHO hats should always be configured and not be left as an obscure setting: here are the files that are working for me right now
[advmameautojs.zip](https://github.com/EmuELEC/EmuELEC/files/8187334/advmameautojs.zip)

it uses advj to determine the hat stick, and it has worked with all the gamepads I have at hand

**Comment by Langerz82** [comment]
Hey, I thought advj is not included in EE. If that works for one, it's ok I can fix it to work with others. I'll use the code and fix it for multiple controllers.

**Comment by shantigilbert** [comment]
It is not, but as you can read in a comment I left earlier I modified it and it will be included now.

**Comment by Langerz82** [comment]
oh sorry I only read the latest comment, my bad haha.

**Comment by Langerz82** [comment]
@shantigilbert 
I can fix it to check for each players controller, the only thing is it will have to assume the hat value is the last axis.

**Comment by shantigilbert** [comment]
> @shantigilbert I can fix it to check for each players controller, the only thing is it will have to assume the hat value is the last axis.

If I understand your script correctly, you are assuming all controllers use the same "stick" (hat) you need to run `advj | grep sticks: | sed "s|sticks:\ ||"` for each gamepad and that will tell you which "stick" is the hat, Because if you have 2 gamepads that have different value for the hat (e.g. Xbox 360 having it at 4 but my Hori gamepad has it at 3) the script will not work.

**Comment by Langerz82** [comment]
@shantigilbert Yep, pretty much, it has to detect each one in set gamepad. Is it ok that I assume the last stick will be the hat value? Otherwise we would have to modify the advj executable.

**Comment by shantigilbert** [comment]
> @shantigilbert Yep, pretty much, it has to detect each one in set gamepad. Is it ok that I assume the last stick will be the hat value? Otherwise we would have to modify the advj executable.

Not sure if you are reading my comments... but like I said (3 times now) I already modified the advj executable and it returns the hat when running this `advj | grep sticks: | sed "s|sticks:\ ||"` whether it works on all gamepads I do not know, but it works fine with all the ones I have tried.

**Comment by Langerz82** [comment]
Sorry I was half asleep. Can you send the compiled advj file so I can test? My EE-build will be slow cause I cleaned the cache.

**Comment by Langerz82** [comment]
oh I see now, the 2nd line in the output is the hat value, ok bro, I will make the change.

**Comment by Langerz82** [comment]
@shantigilbert ok test if that works, it should pickup the hat value for each player.

**Comment by Langerz82** [comment]
hold up, bug, the hat values need to reset on 2nd time.

**Comment by Langerz82** [comment]
Need to test, I will build EE and test on my device.

**Comment by shantigilbert** [comment]
Seems to work fine! :+1:

**Comment by Langerz82** [comment]
hey I needed to fix the cfg setup paths in advmame.sh, and add the bios folder to advmame default rom path. I'm testing my other changes atm.

**Comment by shantigilbert** [comment]
There is one small issue. 

On my horipad select (insert coin) and start get mapped to button 7 and 8 when they should be mapped to button 8 and button 9
but it seems to be because the gamecontrollerdb is not returning the correct config, this is the log:

```
CONTROLLER: 1 js0 03000000d11800000094000011010000 Logitech Gamepad F310
GC_CONFIG=03000000d11800000094000011010000,Google LLC Stadia Controller rev. A,platform:Linux,leftstick:b9,rightx:a2,leftshoulder:b4,start:b7,lefty:a1,dpup:h0.1,righty:a3,a:b1,b:b0,guide:b8,dpdown:h0.4,rightshoulder:b5,righttrigger:b13,rightstick:b10,dpright:h0.2,x:b3,back:b6,leftx:a0,y:b2,dpleft:h0.8,lefttrigger:b14,
CONTROLLER: 2 js1 030000006d0400001dc2000014400000 HORI CO.,LTD. HORIPAD S
GC_CONFIG=030000006d0400001dc2000014400000,Logitech F310 Gamepad (XInput),a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b8,leftshoulder:b4,leftstick:b9,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b10,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,platform:Linux,
```

**Comment by shantigilbert** [comment]
> hey I needed to fix the cfg setup paths in advmame.sh, and add the bios folder to advmame default rom path. I'm testing my other changes atm.

last commit was not necessary, and as you can see it conflicts with what I have, either pull the changes from the emuelec repo or revert that last commit.

**Comment by Langerz82** [comment]
the advmame.sh you updated completely broke my 4.3 version. I had to put those changes in for it to work!

**Review by shantigilbert — CHANGES_REQUESTED** [review]
This last commit is not necessary, you are using an older version of the file and it conflicts with the changes. Please revert this commit and just focus on the gamepad.

**Comment by shantigilbert** [comment]
> the advmame.sh you updated completely broke my 4.3 version. I had to put those changes in for it to work!

Yes but this is NOT for 4.3, this is for 4.4 if you want those changes on 4.3 then keep them in your repo.

**Comment by shantigilbert** [comment]
> There is one small issue.
> 
> On my horipad select (insert coin) and start get mapped to button 7 and 8 when they should be mapped to button 8 and button 9 but it seems to be because the gamecontrollerdb is not returning the correct config, this is the log:
> 
> ```
> CONTROLLER: 1 js0 03000000d11800000094000011010000 Logitech Gamepad F310
> GC_CONFIG=03000000d11800000094000011010000,Google LLC Stadia Controller rev. A,platform:Linux,leftstick:b9,rightx:a2,leftshoulder:b4,start:b7,lefty:a1,dpup:h0.1,righty:a3,a:b1,b:b0,guide:b8,dpdown:h0.4,rightshoulder:b5,righttrigger:b13,rightstick:b10,dpright:h0.2,x:b3,back:b6,leftx:a0,y:b2,dpleft:h0.8,lefttrigger:b14,
> CONTROLLER: 2 js1 030000006d0400001dc2000014400000 HORI CO.,LTD. HORIPAD S
> GC_CONFIG=030000006d0400001dc2000014400000,Logitech F310 Gamepad (XInput),a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b8,leftshoulder:b4,leftstick:b9,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b10,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,platform:Linux,
> ```

I found the issue, you are relaying on es_settings.cfg to get the GUID: 

`local DEVICE_GUID=$(get_es_setting string "INPUT P${y}GUID")` and this should not be like this, since people can change gamepads but forget to change the line in es_settings.cfg You should generate the GUID from the /dev/input and THEN compare it to es_settings.cfg (in case you want to respect the order)

**Comment by Langerz82** [comment]
Hey, sorry I meant a pretty new dev build. The issue is when you have an existing .advance in in /storage and /usr/share/advmame is a file not a dir, it breaks compatibillity updating to the new paths.

I used the get_es_setting for legacy it is not really needed, let me know if you want to remove it.

**Comment by shantigilbert** [comment]
> Hey, sorry I meant a pretty new dev build. The issue is when you have an existing .advance in in /storage and /usr/share/advmame is a file not a dir, it breaks compatibillity updating to the new paths.
> 
> I used the get_es_setting for legacy it is not really needed, let me know if you want to remove it.

I am sorry I am not following...  why would `/usr/share/advmame` be a file? and even if it was a file, why do we care about that file? advmame config is in `/storage/.advance` (which is now a symlink to /emuelec/configs/advmame) anything under `/usr` is read-only and its only used to copy into the `/storage` folder on boot.

Anyways, this PR is getting big and its getting hard to follow as it is, lets focus only on the gamepad, if you feel other aspects of advmame not related to gamepad need fixing, then by all means open another PR or an issue.

**Comment by Langerz82** [comment]
In my dev storage build /usr/share/advmame is the exe of advancemame as it was copied there by a build (maybe from 4.3). But people may upgrade and want to keep there settings.
Your current script breaks cause it's a file.
also I needed to check that the symlink pointed to the new path you set which is why I added the safeguard extra code.
I may be wrong but I think without those extra changes when people upgrade the advmame.sh is going to through errors and it will quit the call to advmame and show up on TvTextviewer.

**Comment by shantigilbert** [comment]
> In my dev storage build /usr/share/advmame is the exe of advancemame as it was copied there by a build (maybe from 4.3). But people may upgrade and want to keep there settings. Your current script breaks cause it's a file. also I needed to check that the symlink pointed to the new path you set which is why I added the safeguard extra code. I may be wrong but I think without those extra changes when people upgrade the advmame.sh is going to throw errors and it will quit the call to advmame and show up on TvTextviewer.

I am sorry, but this is not correct... `/usr/share/advmame` does not exists on EmuELEC, not even on 4.3. (https://github.com/EmuELEC/EmuELEC/blob/dev_4.3/packages/sx05re/emulators/advancemame/package.mk) if it exists on your build then it must have been something you added. But again, please let's just focus on the gamepad, any other issue can be resolved on another PR or issue, let's not mix these.

**Comment by shantigilbert** [comment]
Awesome, it seems it is now working as it should! 

```
BTN_CFG=0 1 2 3 4 5 6 7
CONTROLLER: 1 js0 030000006d0400001dc2000014400000 Logitech Gamepad F310
GC_CONFIG=030000006d0400001dc2000014400000,Logitech F310 Gamepad (XInput),a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b8,leftshoulder:b4,leftstick:b9,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b10,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,platform:Linux,
CONTROLLER: 2 js1 030000000d0f0000c100000011010000 HORI CO.,LTD. HORIPAD S
GC_CONFIG=030000000d0f0000c100000011010000,HORIPAD S,dpleft:h0.8,rightx:a2,dpright:h0.2,rightshoulder:b5,dpdown:h0.4,righty:a3,leftshoulder:b4,y:b0,x:b3,b:b1,a:b2,dpup:h0.1,back:b8,leftstick:b10,start:b9,lefty:a1,guide:b13,lefttrigger:b6,righttrigger:b7,rightstick:b11,leftx:a0,platform:Linux,
```

Thank you! I will merge it now

**Review by shantigilbert — APPROVED** [review]

**Comment by shantigilbert** [comment]
BTW: for future use there is a utility I recently added called `gamepad_info` which you can also use to get the information for gamepads currently connected:

if executed directly it will show the sdl configuration: 

```
# gamepad_info
030000006d0400001dc2000014400000,Logitech F310 Gamepad (XInput),a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b8,leftshoulder:b4,leftstick:b9,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b10,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,platform:Linux,
030000000d0f0000c100000011010000,HORIPAD S,dpleft:h0.8,rightx:a2,dpright:h0.2,rightshoulder:b5,dpdown:h0.4,righty:a3,leftshoulder:b4,y:b0,x:b3,b:b1,a:b2,dpup:h0.1,back:b8,leftstick:b10,start:b9,lefty:a1,guide:b13,lefttrigger:b6,righttrigger:b7,rightstick:b11,leftx:a0,platform:Linux,
```

and if executed with the `-more` argument it will display more information about each gamepad:

```
# gamepad_info -more
Sys_InitInput: Loading /storage/.config/SDL-GameControllerDB/gamecontrollerdb.txt
Opened gamepad device index 0 (Logitech F310 Gamepad (XInput))
        axes: 6
        hats: 1
     buttons: 11
 instance id: 0
        name: Logitech Gamepad F310
        guid: 030000006d0400001dc2000014400000
     mapping: 030000006d0400001dc2000014400000,Logitech F310 Gamepad (XInput),a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b8,leftshoulder:b4,leftstick:b9,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b10,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,platform:Linux,
Opened gamepad device index 1 (HORIPAD S)
        axes: 4
        hats: 1
     buttons: 14
 instance id: 1
        name: HORIPAD S
        guid: 030000000d0f0000c100000011010000
     mapping: 030000000d0f0000c100000011010000,HORIPAD S,dpleft:h0.8,rightx:a2,dpright:h0.2,rightshoulder:b5,dpdown:h0.4,righty:a3,leftshoulder:b4,y:b0,x:b3,b:b1,a:b2,dpup:h0.1,back:b8,leftstick:b10,start:b9,lefty:a1,guide:b13,lefttrigger:b6,righttrigger:b7,rightstick:b11,leftx:a0,platform:Linux,
```

which could be useful to get the information in one single command.

**Comment by Langerz82** [comment]
Ok great. Sounds like it would have made my life much easier, but at least manually coding some stuff gives me practice and expands my knowledge lool.

**Comment by shantigilbert** [comment]
@Langerz82 I have made several changes to clean up the joy_common.sh file. There is only one issue that I can't seem to resolve, maybe you want to take a shot at it. 

https://github.com/EmuELEC/EmuELEC/commit/036c2ebf0a5586ccf506440cc8f8240093c0234c

1.- If the gamepad order is set in es_settings.cfg the gamepads should respect that order
2.- If the gamepad order is NOT set in es_settings.cfg the gamepads should be ordered by js0, js1, js2, etc
3.- If the gamepad order is only partially set in es_settings.cfg the gamepads should be ordered by js0, js1, js2, etc

Right now the order is whatever `gamepad_info` returns, so its based on gamecontrollerdb.txt 

I was thinking sorting the array `$PLAYER_CFG` could work, but I didn't get a chance to try it. 

let me know if you can help

**Comment by shantigilbert** [comment]
I think I got the order, at least by `/dev/input/js?`

**Comment by Langerz82** [comment]
Hey, looks good thanks! Yeah JS index would be the order most of the time, the sdl checks the order properly but is overkill, and probably would not be needed 99% of the time. The only time I think it could be diff is a usb dongle that occupies the start js's but the controller are not on, and a wireless controller that is a separate controller to the dongle connects.
That is the only case I can think of where it differs.

In that case:
Dongle1: js0
Dongle2: js1
Controller3: js2 (rumble) && js3 (Even though it set to player 1).

**Comment by Langerz82** [comment]
Maybe we should just enable the sdl check by default. In the previous case it will mix up the ordering pairs, and users will have screwed up input configs.

**Comment by shantigilbert** [comment]
I pushed a change that sorts it by /dev/input/js* I guess it will do for now.
